### ISSUE TITLE
Moving due delayed messages fails when queue tables contain additional columns

### ISSUE THREAD

**Issue by JamieBenson** [Particular/NServiceBus.SqlServer#1288]
## Symptoms

An exception is thrown:

```log
Exception thrown while moving matured delayed messages
Microsoft.Data.SqlClient.SqlException (0x80131904): Column name or number of supplied values does not match table definition.
```

### Who's affected

Only users with queue tables containing additional columns not created by NServiceBus installers are affected.

### Root cause

The SQL statement to insert delayed messages that are ready to be processed did not specify column names for the insert operation.

## Fixes

* https://github.com/Particular/NServiceBus.SqlServer/pull/1289 for `master`
* https://github.com/Particular/NServiceBus.SqlServer/pull/1290 for `release-7.0`
* https://github.com/Particular/NServiceBus.SqlServer/pull/1291 for `release-6.3`

### Describe the suggested improvement
I’m looking at upgrading components that are combination of V4 and V6 to the latest versions of NServiceBus and NServiceBus SQL Transport versions.

I’ve got a prototype I think is nearly there minus distributed transactions.

However I’m getting the following error from the new native delay functionality.

“Column name or number of supplied values does not match table definition” being sent from the DueDelayProcessor.

This looks to be related to how we have deployed NServiceBus within our company.

#### Is your improvement related to a problem? Please describe.
If the tables have extra default columns, the solution will not work as is.

#### Describe the suggested solution
Update the code to explicit name the inferred columns so the code will be more extensible to how companies deploy NServiceBus.

Example test case was logged on the [discussion portal](https://discuss.particular.net/t/potential-issue-improvement-with-v7-on-sql-transport-with-default-columns/3836).
        
#### Describe alternatives you've considered
Deleting columns off all databases - not ideal and causes standards issues.
Make the SQL configurable somehow - arguably SQL attack vector
Make the SQL scripts extensible somehow, either via subclass, inheritance or interface.

### Additional Context

I logged the information on the [discuss portal](https://discuss.particular.net/t/potential-issue-improvement-with-v7-on-sql-transport-with-default-columns/3836)

**Comment by DavidBoike** [comment]
So, to be clear, the problem you're experiencing is because you've added an extra column `CreatedDate` to all of your queue tables, like this:

```
CREATE TABLE [dbo].[TestTable](
	[Id] [uniqueidentifier] NOT NULL,
	[CorrelationId] [varchar](255) NULL,
	[ReplyToAddress] [varchar](255) NULL,
	[Recoverable] [bit] NOT NULL,
	[Expires] [datetime] NULL,
	[Headers] [varchar](max) NOT NULL,
	[Body] [varbinary](max) NULL,
	[RowVersion] [bigint] IDENTITY(1,1) NOT NULL,
	[CreatedDate] [datetime2](7) NOT NULL
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]
```

Is this the only such extra column that has been added, or are there potentially others?

**Comment by JamieBenson** [comment]
Hi David,

That's correct.  We've got these on all the Input Queue / Audit / Timeout columns etc.

We have other areas that have other default columns (it's a large organisation) and have slightly different implementations based on the DBA / compliance requirements at the time.

I note this may be classed as an unsupported change from your side, but I've worked at many companies that have NServiceBus and make additions to match their compliance policies.

I'm hoping the suggested fix I've supplied is a happy compromise / low risk change that works for us and other consumers of the SQL Transport system.

I should note this worked with V4 and V6, I think it's just when we've upgraded there was an implied implication of what we've done won't work.

Cheers,
Jamie
